### PR TITLE
move: fix delete-empty-src-dirs flag - fixes #2372

### LINKED ISSUE #2372: delete-empty-src-dirs not deleting folders.
Command used: rclone move -vv -L --delete-empty-src-dirs --ignore-existing --fast-list --stats=30s --drive-chunk-size=256M --buffer-size=256M "/home/user1/data/upload/" gdrive:

After uploading everything, every folder inside /home/user1/data/upload/ will still be there.

The log doesn't show anything interesting at the end:

> 2018/07/02 11:34:56 DEBUG : 5 go routines active
> 2018/07/02 11:34:56 DEBUG : rclone: Version "v1.42" finishing with parameters ["rclone" "move" "-vv" "-L" "--delete-empty-src-dirs" "--ignore-existing" "--fast-list" "--stats=30s" "--drive-chunk-size=256M" "--buffer-size=256M" "/home/user1/data/upload/" "gdrive:"

### PR DESCRIPTION
**Pull request by ishuah**
This PR fixes an issue with the `--delete-empty-src-dirs` flag on move where not all empty dirs would not get delete on successful move.

Fixes - #2372

### CODE REVIEW

File: fs/sync/sync.go [hunk 1]
```diff
@@ -522,13 +522,16 @@ func copyEmptyDirectories(f fs.Fs, entries map[string]fs.DirEntry) error {
 	return nil
 }
 
-func parentDirCheck(entries map[string]fs.DirEntry, entry fs.DirEntry) {
+func (s *syncCopyMove) parentDirCheck(entry fs.DirEntry) {
+	if s.DoMove {
```
Comment (ncw) [inline fs/sync/sync.go]: Can you put a commentabove the if - something like

```
	// If we are moving files then we don't want to remove directories with files in
	// from the source as we are about to move them making the directory empty.
```

It took me a while to work  out how your fix worked!

File: fs/sync/sync.go [hunk 2]
```diff
@@ -522,13 +522,16 @@ func copyEmptyDirectories(f fs.Fs, entries map[string]fs.DirEntry) error {
 	return nil
 }
 
-func parentDirCheck(entries map[string]fs.DirEntry, entry fs.DirEntry) {
```
Comment (ncw) [inline fs/sync/sync.go]: The should probably be called `srcParentDirCheck` now it only acts on the source directory.

File: fs/sync/sync_test.go [hunk 3]
```diff
@@ -947,6 +947,28 @@ func testServerSideMove(t *testing.T, r *fstest.Run, withFilter, testDeleteEmpty
...
+// Test move
+func TestMove(t *testing.T) {
+	r := fstest.NewRun(t)
+	defer r.Finalise()
+	file1 := r.WriteFile("sub dir/hello world", "hello world", t1)
+	file2 := r.WriteFile("nested/sub dir/file", "nested", t1)
+	r.Mkdir(r.Fremote)
+
+	// run move with --delete-empty-src-dirs
+	err := MoveDir(r.Fremote, r.Flocal, true)
```
Comment (ncw) [inline fs/sync/sync_test.go]: I'd quite like to see another test without `--delete-empty-src-dirs`, something like this which I wrote convincing myself that your patch worked :-)

```diff
diff --git a/fs/sync/sync_test.go b/fs/sync/sync_test.go
index 57b62d5a..d05acd9d 100644
--- a/fs/sync/sync_test.go
+++ b/fs/sync/sync_test.go
@@ -948,7 +948,7 @@ func testServerSideMove(t *testing.T, r *fstest.Run, withFilter, testDeleteEmpty
 }
 
 // Test move
-func TestMove(t *testing.T) {
+func TestMoveWithDeleteEmptySrcDirs(t *testing.T) {
 	r := fstest.NewRun(t)
 	defer r.Finalise()
 	file1 := r.WriteFile("sub dir/hello world", "hello world", t1)
@@ -969,6 +969,32 @@ func TestMove(t *testing.T) {
 	fstest.CheckItems(t, r.Fremote, file1, file2)
 }
 
+// Test move
+func TestMoveWithoutDeleteEmptySrcDirs(t *testing.T) {
+	r := fstest.NewRun(t)
+	defer r.Finalise()
+	file1 := r.WriteFile("sub dir/hello world", "hello world", t1)
+	file2 := r.WriteFile("nested/sub dir/file", "nested", t1)
+	r.Mkdir(r.Fremote)
+
+	// run move without --delete-empty-src-dirs
+	err := MoveDir(r.Fremote, r.Flocal, false)
+	require.NoError(t, err)
+
+	fstest.CheckListingWithPrecision(
+		t,
+		r.Flocal,
+		nil,
+		[]string{
+			"sub dir",
+			"nested",
+			"nested/sub dir",
+		},
+		fs.GetModifyWindow(r.Flocal),
+	)
+	fstest.CheckItems(t, r.Fremote, file1, file2)
+}
+
 // Test a server side move if possible, or the backup path if not
 func TestServerSideMove(t *testing.T) {
 	r := fstest.NewRun(t)
```

**Comment by ishuah** [comment]
Thanks @ncw! Changes made

**Comment by ncw** [comment]
That looks great - thanks @ishuah :-)
